internal/attestation/signer/signer.go

```diff
@@ -0,0 +1,225 @@
+//
+// Copyright 2024 The Chainloop Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package signer
+
+import (
+	"context"
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"sync"
+	"syscall"
+
+	pb "github.com/chainloop-dev/chainloop/app/controlplane/api/controlplane/v1"
+	"github.com/rs/zerolog"
+	"github.com/sigstore/cosign/v2/pkg/signature"
+	sigstoresigner "github.com/sigstore/sigstore/pkg/signature"
+	sigdsee "github.com/sigstore/sigstore/pkg/signature/dsse"
+	"golang.org/x/term"
+)
+
+// ChainloopSigner is a wrapper of a Sigstore signer for Chainloop
+type ChainloopSigner struct {
+	sigstoresigner.Signer
+
+	keyPath              string
+	signingServiceClient pb.SigningServiceClient
+	logger               zerolog.Logger
+	mu                   sync.Mutex
+}
+
+var _ sigstoresigner.Signer = (*ChainloopSigner)(nil)
+
+func NewChainloopSigner(keyPath string, sc pb.SigningServiceClient, logger zerolog.Logger) *ChainloopSigner {
+	return &ChainloopSigner{
+		keyPath:              keyPath,
+		signingServiceClient: sc,
+		logger:               logger,
+	}
+}
+
+func (cs *ChainloopSigner) SignMessage(message io.Reader, opts ...sigstoresigner.SignOption) ([]byte, error) {
+	err := cs.ensureInitiated(context.Background())
+	if err != nil {
+		return nil, err
+	}
+
+	return cs.Signer.SignMessage(message, opts...)
+}
+
+func (cs *ChainloopSigner) ensureInitiated(ctx context.Context) error {
```

Comment:
can you document a little bit that this will pick one or the other and that it mutates the signer, it's not that clear.